### ISSUE TITLE
POST with empty `Content-Type` generates an error

### ISSUE THREAD

**Issue by authchir** [jooby-project/jooby#444]
Hello,
Given the following minimal setup:

```Java
import java.nio.*;
import java.nio.file.*;
import org.jooby.*;

public class Main extends Jooby {
  {
    parser((type, ctxt) -> {
      if (type.getRawType() == ByteBuffer.class) {
        return ctxt.body(body -> {
          return ByteBuffer.wrap(body.bytes());
        });
      }
      return ctxt.next();
    });

    post("/upload", req -> {
      ByteBuffer buffer = req.body().to(ByteBuffer.class);
      Files.write(Paths.get("upload"), buffer.array());
      return "";
    });
  }

  public static void main(final String[] args) throws Throwable {
    new Main().start(args);
  }
}
```

The following works correctly:
```
$ curl -i --request POST --header 'Content-Type: text/plain' --data 'abc' localhost:8080/upload
HTTP/1.1 200 OK
Content-Type: text/html;charset=UTF-8
Content-Length: 0
connection: keep-alive
```
While the following generates an error:
```
$ curl -i --request POST --data 'abc' localhost:8080/upload
HTTP/1.1 400 Bad Request
Cache-Control: must-revalidate,no-cache,no-store
Content-Type: text/html;charset=UTF-8
Content-Length: 2834
connection: keep-alive

<!doctype html>
<html>
<head>
<meta charset="UTF-8">
<style>
body {font-family: "open sans",sans-serif; margin-left: 20px;}
h1 {font-weight: 300; line-height: 44px; margin: 25px 0 0 0;}
h2 {font-size: 16px;font-weight: 300; line-height: 44px; margin: 0;}
footer {font-weight: 300; line-height: 44px; margin-top: 10px;}
hr {background-color: #f7f7f9;}
div.trace {border:1px solid #e1e1e8; background-color: #f7f7f9;}
p {padding-left: 20px;}
p.tab {padding-left: 40px;}
</style>
<title>
400 Bad Request
</title>
<body>
<h1>Bad Request</h1>
<hr><h2>message: Bad Request(400): Failed to parse parameters to 'java.nio.ByteBuffer'</h2>
<h2>status: 400</h2>
<h2>stack:</h2>
<div class="trace">
<p class="line"><code>org.jooby.Err: Bad Request(400): Failed to parse parameters to 'java.nio.ByteBuffer'</code></p>
<p class="line tab"><code>  at org.jooby.internal.MutantImpl.to(MutantImpl.java:87)</code></p>
<p class="line tab"><code>  at org.jooby.internal.MutantImpl.to(MutantImpl.java:75)</code></p>
<p class="line tab"><code>  at org.jooby.Mutant.to(Mutant.java:307)</code></p>
<p class="line tab"><code>  at com.gfaga.portail.Main.lambda$new$2(Main.java:27)</code></p>
<p class="line tab"><code>  at org.jooby.Route$OneArgHandler.handle(Route.java:1727)</code></p>
<p class="line tab"><code>  at org.jooby.internal.RouteImpl.handle(RouteImpl.java:108)</code></p>
<p class="line tab"><code>  at org.jooby.internal.RouteChain.next(RouteChain.java:70)</code></p>
<p class="line tab"><code>  at org.jooby.Route$Chain.next(Route.java:2074)</code></p>
<p class="line tab"><code>  at org.jooby.internal.HttpHandlerImpl.handle(HttpHandlerImpl.java:293)</code></p>
<p class="line tab"><code>  at org.jooby.internal.netty.NettyHandler.channelRead0(NettyHandler.java:85)</code></p>
<p class="line tab"><code>  at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:105)</code></p>
<p class="line tab"><code>  at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:372)</code></p>
<p class="line tab"><code>  at io.netty.channel.AbstractChannelHandlerContext.access$600(AbstractChannelHandlerContext.java:38)</code></p>
<p class="line tab"><code>  at io.netty.channel.AbstractChannelHandlerContext$7.run(AbstractChannelHandlerContext.java:363)</code></p>
<p class="line tab"><code>  at io.netty.util.concurrent.DefaultEventExecutor.run(DefaultEventExecutor.java:66)</code></p>
<p class="line tab"><code>  at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:877)</code></p>
<p class="line tab"><code>  at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:144)</code></p>
<p class="line tab"><code>  at java.lang.Thread.run(Thread.java:745)</code></p>
</div>
</body>
</html>
```

According to [*§3.1.1.5 Content-Type*][1] of RFC 7231:
```
A sender that generates a message containing a payload body SHOULD
generate a Content-Type header field in that message unless the
intended media type of the enclosed representation is unknown to the
sender.  If a Content-Type header field is not present, the recipient
MAY either assume a media type of "application/octet-stream"
([RFC2046], Section 4.5.1) or examine the data to determine its type.
```

I tried a little bit to track down the origin of this behaviour and could only confirm, that the parser is successfully added to the context, but for some reason it is not being used.

[1]: https://tools.ietf.org/html/rfc7231#section-3.1.1.5

**Comment by jknack** [comment]
will look at this soon... but have two questions/suggestions

1. how do you find the Parser API? is it complex?

2. I suggested you to start the app with the run method:

`run(Main::new, args)`

It deals with some logback issues (you probably didn't find one yet) but is the recommended approach

**Comment by authchir** [comment]
Hello,
I find it quite simple to use the API and I appreciate the flexibility one has to handle a specific situation (e.g. I could read the first bytes of the body to check for a specific version of a file format). I find it way more interesting and flexible than the alternative `Map<Type, Function>`.

However, I am not convinced of the *call the next* strategy. It means a programming mistake in one parser could prevent the ones following in the list to be executed. I personally think this should be a responsibility of the underlying infrastructure (e.g. the parser could return an `Optional<Object>`: `Optional.of(...)` if the parser can handle the request and `Optional.none()` otherwise).

What are your thoughts on that?

**Comment by jknack** [comment]
Thought on that too.. but get tricky with container types like: `List`, `Set` or even `Optional`:

Collection parser works like:

`
List list = ...;
for (Object value : values) {
   list.add(ctx.next(paramType, value));
}
return list;
`

**Comment by authchir** [comment]
The problem came back in an other form:

```
$ curl -i --request POST --data 'abc' localhost:8080/upload                                                                                    
HTTP/1.1 500 Internal Server Error
Cache-Control: must-revalidate,no-cache,no-store
Content-Type: text/html;charset=UTF-8
Content-Length: 7845
connection: keep-alive

<!doctype html>
<html>
<head>
<meta charset="UTF-8">
<style>
body {font-family: "open sans",sans-serif; margin-left: 20px;}
h1 {font-weight: 300; line-height: 44px; margin: 25px 0 0 0;}                                                                                                       [39/1698]
h2 {font-size: 16px;font-weight: 300; line-height: 44px; margin: 0;}
footer {font-weight: 300; line-height: 44px; margin-top: 10px;}
hr {background-color: #f7f7f9;}
div.trace {border:1px solid #e1e1e8; background-color: #f7f7f9;}
p {padding-left: 20px;}
p.tab {padding-left: 40px;}
</style>
<title>
500 Server Error
</title>
<body>
<h1>Server Error</h1>
<hr><h2>message: Ambigous constructor found: java.nio.ByteBuffer. Bean/Form type must have a no-args constructor or must be annotated with @javax.inject.Inject</h2>
<h2>status: 500</h2>
<h2>stack:</h2>
<div class="trace">
<p class="line"><code>java.lang.IllegalStateException: Ambigous constructor found: java.nio.ByteBuffer. Bean/Form type must have a no-args constructor or must be annotated w
ith @javax.inject.Inject</code></p>
<p class="line tab"><code>  at org.jooby.internal.parser.bean.BeanPlan.<init>(BeanPlan.java:95)</code></p>
<p class="line tab"><code>  at org.jooby.internal.parser.BeanParser.newBean(BeanParser.java:90)</code></p>
<p class="line tab"><code>  at org.jooby.internal.parser.BeanParser.lambda$parse$1(BeanParser.java:73)</code></p>
<p class="line tab"><code>  at org.jooby.internal.parser.ParserBuilder.parse(ParserBuilder.java:116)</code></p>
<p class="line tab"><code>  at org.jooby.internal.parser.ParserExecutor$1.next(ParserExecutor.java:161)</code></p>
<p class="line tab"><code>  at org.jooby.internal.parser.ParserBuilder.parse(ParserBuilder.java:114)</code></p>
<p class="line tab"><code>  at org.jooby.internal.parser.ParserExecutor$1.next(ParserExecutor.java:161)</code></p>
<p class="line tab"><code>  at org.jooby.internal.parser.ParserBuilder.parse(ParserBuilder.java:114)</code></p>
<p class="line tab"><code>  at org.jooby.internal.parser.ParserExecutor$1.next(ParserExecutor.java:161)</code></p>
<p class="line tab"><code>  at org.jooby.internal.parser.ParserBuilder.parse(ParserBuilder.java:114)</code></p>
<p class="line tab"><code>  at org.jooby.internal.parser.ParserExecutor$1.next(ParserExecutor.java:161)</code></p>
<p class="line tab"><code>  at org.jooby.internal.parser.ParserBuilder.parse(ParserBuilder.java:114)</code></p>
<p class="line tab"><code>  at org.jooby.internal.parser.ParserExecutor$1.next(ParserExecutor.java:161)</code></p>
<p class="line tab"><code>  at org.jooby.internal.parser.ParserExecutor$1.next(ParserExecutor.java:135)</code></p>
<p class="line tab"><code>  at org.jooby.internal.parser.LocaleParser.parse(LocaleParser.java:37)</code></p>
<p class="line tab"><code>  at org.jooby.internal.parser.ParserExecutor$1.next(ParserExecutor.java:158)</code></p>
<p class="line tab"><code>  at org.jooby.internal.parser.ParserExecutor$1.next(ParserExecutor.java:135)</code></p>
<p class="line tab"><code>  at org.jooby.internal.parser.LocalDateParser.parse(LocalDateParser.java:52)</code></p>
<p class="line tab"><code>  at org.jooby.internal.parser.ParserExecutor$1.next(ParserExecutor.java:158)</code></p>
<p class="line tab"><code>  at org.jooby.internal.parser.ParserExecutor$1.next(ParserExecutor.java:135)</code></p>
<p class="line tab"><code>  at org.jooby.internal.parser.DateParser.parse(DateParser.java:45)</code></p>
<p class="line tab"><code>  at org.jooby.internal.parser.ParserExecutor$1.next(ParserExecutor.java:158)</code></p>
<p class="line tab"><code>  at org.jooby.internal.parser.ParserBuilder.parse(ParserBuilder.java:114)</code></p>
<p class="line tab"><code>  at org.jooby.internal.parser.ParserExecutor$1.next(ParserExecutor.java:161)</code></p>
<p class="line tab"><code>  at org.jooby.internal.parser.ParserExecutor$1.next(ParserExecutor.java:135)</code></p>
<p class="line tab"><code>  at org.jooby.internal.BuiltinParser$6.parse(BuiltinParser.java:225)</code></p>
<p class="line tab"><code>  at org.jooby.internal.parser.ParserExecutor$1.next(ParserExecutor.java:158)</code></p>
<p class="line tab"><code>  at org.jooby.internal.parser.ParserExecutor$1.next(ParserExecutor.java:135)</code></p>
<p class="line tab"><code>  at org.jooby.internal.BuiltinParser$5.parse(BuiltinParser.java:214)</code></p>
<p class="line tab"><code>  at org.jooby.internal.parser.ParserExecutor$1.next(ParserExecutor.java:158)</code></p>
<p class="line tab"><code>  at org.jooby.internal.parser.ParserExecutor$1.next(ParserExecutor.java:135)</code></p>
<p class="line tab"><code>  at org.jooby.internal.BuiltinParser$4.parse(BuiltinParser.java:203)</code></p>
<p class="line tab"><code>  at org.jooby.internal.parser.ParserExecutor$1.next(ParserExecutor.java:158)</code></p>
<p class="line tab"><code>  at org.jooby.internal.parser.ParserExecutor$1.next(ParserExecutor.java:135)</code></p>
<p class="line tab"><code>  at org.jooby.internal.BuiltinParser$3.parse(BuiltinParser.java:185)</code></p>
<p class="line tab"><code>  at org.jooby.internal.parser.ParserExecutor$1.next(ParserExecutor.java:158)</code></p>
<p class="line tab"><code>  at org.jooby.internal.parser.ParserExecutor$1.next(ParserExecutor.java:135)</code></p>
<p class="line tab"><code>  at org.jooby.internal.BuiltinParser$2.parse(BuiltinParser.java:154)</code></p>
<p class="line tab"><code>  at org.jooby.internal.parser.ParserExecutor$1.next(ParserExecutor.java:158)</code></p>
<p class="line tab"><code>  at org.jooby.internal.parser.ParserExecutor$1.next(ParserExecutor.java:135)</code></p>
<p class="line tab"><code>  at org.jooby.internal.BuiltinParser$1.parse(BuiltinParser.java:84)</code></p>
<p class="line tab"><code>  at org.jooby.internal.parser.ParserExecutor$1.next(ParserExecutor.java:158)</code></p>
<p class="line tab"><code>  at org.jooby.internal.parser.ParserExecutor.convert(ParserExecutor.java:74)</code></p>
<p class="line tab"><code>  at org.jooby.internal.MutantImpl.to(MutantImpl.java:85)</code></p>
<p class="line tab"><code>  at org.jooby.internal.MutantImpl.to(MutantImpl.java:76)</code></p>
<p class="line tab"><code>  at org.jooby.Mutant.to(Mutant.java:307)</code></p>
<p class="line tab"><code>  at com.gfaga.portail.Main.lambda$new$2(Main.java:34)</code></p>
<p class="line tab"><code>  at org.jooby.Route$OneArgHandler.handle(Route.java:1727)</code></p>
<p class="line tab"><code>  at org.jooby.internal.RouteImpl.handle(RouteImpl.java:108)</code></p>
<p class="line tab"><code>  at org.jooby.internal.RouteChain.next(RouteChain.java:70)</code></p>
<p class="line tab"><code>  at org.jooby.Route$Chain.next(Route.java:2074)</code></p>
<p class="line tab"><code>  at org.jooby.internal.HttpHandlerImpl.handle(HttpHandlerImpl.java:304)</code></p>
<p class="line tab"><code>  at org.jooby.internal.netty.NettyHandler.channelRead0(NettyHandler.java:92)</code></p>
<p class="line tab"><code>  at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:105)</code></p>
<p class="line tab"><code>  at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:372)</code></p>
<p class="line tab"><code>  at io.netty.channel.AbstractChannelHandlerContext.access$600(AbstractChannelHandlerContext.java:38)</code></p>
<p class="line tab"><code>  at io.netty.channel.AbstractChannelHandlerContext$7.run(AbstractChannelHandlerContext.java:363)</code></p>
<p class="line tab"><code>  at io.netty.util.concurrent.DefaultEventExecutor.run(DefaultEventExecutor.java:66)</code></p>
<p class="line tab"><code>  at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:873)</code></p>
<p class="line tab"><code>  at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:144)</code></p>
<p class="line tab"><code>  at java.lang.Thread.run(Thread.java:745)</code></p>
</div>
</body>
</html>
```

**Comment by jknack** [comment]
damn it! just released 1.0.0.CR8

will have a look later

**Comment by jknack** [comment]
Mixed feelings here, this call: 

    curl -i --request POST --data 'abc' localhost:8080/upload

Give us an `application/x-www-form-urlencoded` content type, that is why jooby try to parse body as JavaBean.

Having said that, think here it should do what you ask to do: `Get a ByteBuffer`

Think I'm going to resolve this by:

1. removing a form post parsing from: `body()` method
2. people who want a form post should use `params()` method or `form(Class)` method
3. add a `form(Class)` method which is just an alias for `params(Class)`
